### PR TITLE
ARROW-9630: [Go] Implement public JSON reader/writer

### PR DESCRIPTION
**Pull request by zeroshade**
This JSON reader/writer is intended to support the standard Go encoding/json package for Go and writes out arrays and records as data, not as the integration format. This implements `MarshalJSON` for the array types, and `UnmarshalJSON` for the array builders (in order to unmarshal json data to add to an existing builder) along with providing a few helper functions for single array from json or single recordbatch from json logic.

### CODE REVIEW

File: go/arrow/array/binary.go
Comment (emkornfield) [inline go/arrow/array/binary.go:133]: How do non-printable characters get marshalled in json?  Escaped bytes?

Comment (emkornfield) [reply go/arrow/array/binary.go:133]: Okay, based on below is looks like base64 encoded?  This is not something I could find as a reference in the docs?

File: go/arrow/array/decimal128.go
Comment (emkornfield) [inline go/arrow/array/decimal128.go:268]: any strong rationale for ToNearestAway?  maybe document?

Comment (emkornfield) [inline go/arrow/array/decimal128.go:268]: any strong rationale for ToNearestAway?  maybe document?

File: go/arrow/array/binary.go
Comment (zeroshade) [reply go/arrow/array/binary.go:133]: From the Go docs on JSON marshalling: https://pkg.go.dev/encoding/json#Marshal

> Array and slice values encode as JSON arrays, except that []byte encodes as a base64-encoded string, and a nil slice encodes as the null JSON value.

The standard marshalling of byte slices to JSON in go will marshal it to a base64-encoded string.

File: go/arrow/array/decimal128.go
Comment (zeroshade) [reply go/arrow/array/decimal128.go:268]: No strong rationale, just happened to be what got me the closest to the value I had input when I used it and I couldn't think of a good way to easily make this configurable in the current interface. I'll add a comment to document it.

Comment (zeroshade) [reply go/arrow/array/decimal128.go:268]: duplicate comment of https://github.com/apache/arrow/pull/11359#discussion_r741359958

Comment (emkornfield) [reply go/arrow/array/decimal128.go:268]: Also, is 0 the right value for precision here shouldn't it be 128?

Also if I read the docs correctly it seems that this will silently truncate decimal values that require larger precision, do you have thoughts on if it would be better to check for these types of values and return an error instead?

File: go/arrow/array/decimal128.go
Comment (emkornfield) [inline go/arrow/array/decimal128.go:272]: I might be misunderstanding but it seems like json.Number is an alias for string (https://pkg.go.dev/encoding/json#Number).  Should this go through the same logic above for string?

File: go/arrow/array/fixed_size_list.go
Comment (emkornfield) [inline go/arrow/array/fixed_size_list.go:133]: Is this using a string to encode FixedSizeLists?  That might be counter-intuitive to people?  I guess this is needed to round-trip with full-fidelity?

File: go/arrow/array/float16_builder.go
Comment (emkornfield) [inline go/arrow/array/float16_builder.go:186]: Does this silently truncate or does it throw an error?

File: go/arrow/array/float16_builder.go
Comment (emkornfield) [inline go/arrow/array/float16_builder.go:222]: check for copy and paste but in error messages?

File: go/arrow/array/interval.go
Comment (emkornfield) [inline go/arrow/array/interval.go:105]: In this case, it seems like it might be better to have a formatted string (like "1m") or "1months"

File: go/arrow/array/interval.go
Comment (emkornfield) [inline go/arrow/array/interval.go:353]: same comment, as months, consider making a formatted string that can be reparsed?  I guess whether we use a formatted string or the raw values comes down to who we expect to consume the JSON.  I think for machines the numbers might be more natural, but for human consumption formatted string makes it clearer what the types are.

File: go/arrow/array/map.go
Comment (emkornfield) [inline go/arrow/array/map.go:289]: copy-paste error message?

File: go/arrow/array/numeric.gen.go
Comment (emkornfield) [inline go/arrow/array/numeric.gen.go:90]: won't this truncate int64 values that require > 52 bits of storage?

File: go/arrow/array/decimal128.go
Comment (zeroshade) [reply go/arrow/array/decimal128.go:272]: It's not an alias but an actual type, which would only get used if the json encoder sets an option to use it. The `json.Number` type provides helper functions for parsing from the string, so the logic wouldn't be the same as using string, but instead using the helper functions directly and checking the error returns from them

File: go/arrow/array/numeric.gen.go
Comment (emkornfield) [inline go/arrow/array/numeric.gen.go:969]: should formatting be conditional on precision?

File: go/arrow/array/numeric.gen.go
Comment (emkornfield) [inline go/arrow/array/numeric.gen.go:1052]: since it is a different type, it is probably easy to format this to only the precision required?

File: go/arrow/array/fixed_size_list.go
Comment (zeroshade) [reply go/arrow/array/fixed_size_list.go:133]: No, it's just encoding a fixed size list identically to how it encodes regular lists, as a nested list. From the perspective of JSON there's no difference between a `List` where all of the lists happen to be the same length, and a `FixedSizeList`. A fixed size list with `n=3` of Int would encode in json as `[ [1,2,3], [4,5,6], [7,8,9] ]`

File: go/arrow/array/float16_builder.go
Comment (zeroshade) [reply go/arrow/array/float16_builder.go:186]: Currently it will silently truncate

File: go/arrow/array/numeric.gen.go
Comment (emkornfield) [inline go/arrow/array/numeric.gen.go:1384]: this logic is a little opaque to me, would you mind adding a comment on intended output.

Comment (emkornfield) [reply go/arrow/array/numeric.gen.go:1384]: Are there overflow issues for courser precisions?

File: go/arrow/array/struct.go
Comment (emkornfield) [inline go/arrow/array/struct.go:118]: I don't think the arrow spec says struct names need to be unique, it might be worth checking for that case here and throwing an error OR using a list of tuple representation in that case.

File: go/arrow/array/interval.go
Comment (zeroshade) [reply go/arrow/array/interval.go:353]: Currently DayTime interval would get marshaled as `{ "days": #, "milliseconds": # }`, the same would be true for month day nano interval type being marshalled as `{ "months": #, "days": #, "nanoseconds": # }`. The only one that currently wouldn't get marshalled into a formatted object or string is the month interval. The intention I had was that the actual types should be defined externally before attempting to parse the JSON as I didn't build any arrow type inference into the JSON reader. It requires being passed a schema or data type.

File: go/arrow/array/numeric.gen.go
Comment (zeroshade) [reply go/arrow/array/numeric.gen.go:90]: Fair, I'll update the template to only add that to the uint8 type

File: go/arrow/array/struct.go
Comment (emkornfield) [inline go/arrow/array/struct.go:128]: I might be getting a little lost here but IIUC the other json functions return a list representing the column value.  Here it seems we transpose each column to a row first?  This seems to lack symmetry?  Maybe in this case the MarshalJSON function shouldn't use getOneForMarshal?

File: go/arrow/array/util.go
Comment (emkornfield) [inline go/arrow/array/util.go:73]: Maybe also document time/interal/duration types?

File: go/arrow/array/util_test.go
Comment (emkornfield) [inline go/arrow/array/util_test.go:309]: oh is the formatter smart to not extend 0 values on the decimal?

File: go/arrow/datatype_fixedwidth.go
Comment (emkornfield) [inline go/arrow/datatype_fixedwidth.go:113]: I know I asked this before but Go panics on overflow?

Comment (emkornfield) [inline go/arrow/datatype_fixedwidth.go:113]: I think at least in C++ errors are raised when if precision is truncated (i.e. we'd expect everything past the millisecond precision to be 0)

**Review by emkornfield — CHANGES_REQUESTED** [review]
I think this will be really useful.  A few of the comments I left (JSON marshalling of struct type) and behavior on truncation are a little philosophical, I think for data-processing systems being conservative and by default erroring on truncation can be less surprising for users later  on when they try to figure out why precision is missing, but it is in some ways less userfriendly for something as free-form as JSON.  It doesn't seem like there is a great way to control the behavior through the provided go JSON marshaling APIs so we need to pick one, either way we should clearly document the choice and see if this is maybe controllable at the Reader level

File: go/arrow/array/numeric.gen.go
Comment (zeroshade) [reply go/arrow/array/numeric.gen.go:969]: using `9` there will leave out trailing zeros from the resulting string so this will handle whatever level of precision is provided rather than explicitly padding to a specific precision level.

File: go/arrow/array/numeric.gen.go
Comment (zeroshade) [reply go/arrow/array/numeric.gen.go:1052]: using the `9`'s there will drop any trailing zeros so that it will only format it to the precision required to represent the value.

File: go/arrow/array/struct.go
Comment (zeroshade) [reply go/arrow/array/struct.go:118]: currently the Go implementation does not support duplicate field names in structs, only in schemas. So at least for now i'm guaranteed that there will be no duplicate field names until i make a change to support them in the struct type in general.

File: go/arrow/array/struct.go
Comment (zeroshade) [reply go/arrow/array/struct.go:128]: In a struct column, each "value" in the list, is an object consisting of keys from the struct field names with values for the corresponding index. 

For example, a `struct<a: int, b: float>` column would be marshalled to `[ {"a": <a[0]>, "b": <b[0]>}, {"a": <a[1]>, "b": <b[1]> }, .... ]`. Does that make more sense?

File: go/arrow/array/struct.go
Comment (emkornfield) [reply go/arrow/array/struct.go:118]: hmm, OK, I might be wrong about the struct then, I might have been thinking about schema.

File: go/arrow/array/util_test.go
Comment (zeroshade) [reply go/arrow/array/util_test.go:309]: Right, that's the difference between using 0's or 9's in the format string. Using 0's will extend it to the exact precision stated, using 9's will truncate trailing zeros.

File: go/arrow/array/fixed_size_list.go
Comment (zeroshade) [reply go/arrow/array/fixed_size_list.go:133]: I can understand the confusion now that I think about it. The return from `MarshalJSON` is the byte slice of the raw JSON that is produced by the conversion. I hope that clears things up a bit.

File: go/arrow/array/numeric.gen.go
Comment (zeroshade) [reply go/arrow/array/numeric.gen.go:1384]: Yes, there can end up being overflow issues for large enough values depending on the precision. I'll see if i can tweak this to try to avoid the overflows.

File: go/arrow/datatype_fixedwidth.go
Comment (zeroshade) [reply go/arrow/datatype_fixedwidth.go:113]: Looking at https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/value_parsing.h#L530 it looks like the C++ doesn't even accept the excess 0s. It errors if more than the right number of decimal places are provided. So i should be able to easily handle the error cases by using the unit and checking the length of the string. I'll add the errors and update the tests.

File: go/arrow/array/decimal128.go
Comment (zeroshade) [reply go/arrow/array/decimal128.go:272]: comments added

File: go/arrow/array/float16_builder.go
Comment (zeroshade) [reply go/arrow/array/float16_builder.go:187]: comment added to document the truncation

File: go/arrow/array/numeric.gen.go
Comment (zeroshade) [reply go/arrow/array/numeric.gen.go]: fixed

File: go/arrow/array/numeric.gen.go
Comment (zeroshade) [reply go/arrow/array/numeric.gen.go]: fixed

File: go/arrow/array/util.go
Comment (zeroshade) [reply go/arrow/array/util.go:73]: added documentation for those types

**Comment by zeroshade** [comment]
@emkornfield I've addressed the comments you left either via comments or via changes. I tend to lend more towards being less unfriendly wherever possible rather than letting it fail personally which is why I took the tact I did in a lot of the places, but I also see the value in matching behavior of the C++ library in a lot of this since there isn't really a published "spec" for the a JSON form of Arrow it's hard to standardize. Let me know what you think of the changes I've made and comments/documentation I've added.

File: go/arrow/array/decimal128.go
Comment (emkornfield) [reply go/arrow/array/decimal128.go:276]: But it like accessing the [string literal](https://pkg.go.dev/encoding/json#Number.String) is supposed.  It seems like Float64 decoding loses precision needlessly

File: go/arrow/array/float16_builder.go
Comment (emkornfield) [reply go/arrow/array/float16_builder.go:187]: OK, we should probably also make this clear in public docs.

File: go/arrow/array/interval.go
Comment (emkornfield) [reply go/arrow/array/interval.go:105]: sorry, not seeing this appear (maybe forgot to push?)

File: go/arrow/array/interval.go
Comment (emkornfield) [reply go/arrow/array/interval.go:353]: SGTM, might be worth documenting what the value looks like here.

File: go/arrow/array/struct.go
Comment (emkornfield) [reply go/arrow/array/struct.go:128]: Yeah, so I think I've come around to how this is implemented.  I think what I was struggling with is an IntArray, gets serialized as `[1,2,3]` but one could maybe think it should get serialized as something like `[[1],[2],[3]` or `[{'entry':1}, {'entry':2}, {'entry': 3]` similar to what struct does).  I guess it is good if someone just wants to serialize/deserialize just an array and not a record batch that it happens as a flat JSON list and the row by row approach is taken for struct.

**Comment by emkornfield** [comment]
> I've addressed the comments you left either via comments or via changes. I tend to lend more towards being less unfriendly wherever possible rather than letting it fail personally which is why I took the tact I did in a lot of the places, but I also see the value in matching behavior of the C++

It's not so much about matching C++, as it is about users getting unexpected results.  I don't know if there is a good way to disclose some of caveats in a way a user will actually look at before blindly going ahead and using the library.  For now clear documentation for both reading and writing that data is attempted to be coerced as best as possible to the schema provided and might lose some data from the original JSON is fine.  We can re-assess if anyone opens up bugs because of this.

File: go/arrow/array/interval.go
Comment (zeroshade) [reply go/arrow/array/interval.go:353]: Just to add on since my last comment, I've since modified the Month Interval to marshal as a string like you suggested as "1months" etc. but sure, i'll add a specific comment on these marshal functions about their format so it's not only on the top level "to json" functions.

File: go/arrow/array/interval.go
Comment (zeroshade) [reply go/arrow/array/interval.go:105]: It was handled by the change in the `datatype_fixedwidth.go` file by adding `MarshalJSON` and `UnmarshalJSON` functions to the `MonthInterval` type here: https://github.com/apache/arrow/pull/11359/files#diff-ada6f476fe71c04e54cb986f027d91bad2517d3a5a1f278115a80ea3bcc18482R333 When it goes to marshal the array of `MonthInterval` values, it'll use the `MarshalJSON` and `UnmarshalJSON` functions defined on the type, rather than defaulting to just treating it like an integer.

File: go/arrow/array/float16_builder.go
Comment (zeroshade) [reply go/arrow/array/float16_builder.go:187]: It's in the docstring for the `UnmarshalJSON` function which means it'll appear in the public docs for the `UnmarshalJSON` on `Float16Builder`

File: go/arrow/array/decimal128.go
Comment (zeroshade) [reply go/arrow/array/decimal128.go:276]: good point, i've updated the `json.Number` case to work the same as the `string` case and fixed the precision to be 128

File: go/arrow/array/interval.go
Comment (zeroshade) [reply go/arrow/array/interval.go:361]: comments added.

**Comment by zeroshade** [comment]
@emkornfield I've updated based on the feedback. I'll wait until I get an approval from you to merge unless you have more comments or something i missed. Thanks again for the review!

**Comment by emkornfield** [comment]
Thanks, I'll review the updates.  It occurred to me that for Interval months it should maybe serialize as `{"months": X}` which would allow import as months and also potentially as `MonthDayNano`.  Sorry for the churn here and feel free for ignore this.

File: go/arrow/array/interval.go
Comment (emkornfield) [inline go/arrow/array/interval.go:98]: not sure if you saw my last comment what do you think about thins`{months=num}`

**Comment by emkornfield** [comment]
LGTM modulo the months interval comment which I don't feel too strongly about.

File: go/arrow/array/interval.go
Comment (zeroshade) [reply go/arrow/array/interval.go:98]: I saw the comment, I wasn't feeling too strongly one way or the other but sure, I'll bump it to that format since it's easy enough to do. After everything passes I can commit this, if you're willing to set your review to approve the changes :)

Comment (zeroshade) [reply go/arrow/array/interval.go:98]: latest commit changes the format to `{"months": #}` and I updated the comments to reflect that in the documentation.

**Review by emkornfield — APPROVED** [review]
I think you need to rebase.

**Comment by zeroshade** [comment]
@emkornfield  Yup, just did the rebase. Once everything passes i'll merge. Thanks for the review and comments!

**Pull request closed without merging**
